File: GoBackN.java
import java.util.*;

public class GoBackN {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        System.out.print("Total Frames: "); int total = sc.nextInt();
        System.out.print("Window Size: ");  int w = sc.nextInt();

        int base = 0;
        Random rnd = new Random();

        while (base < total) {
            int end = Math.min(base + w, total);
            System.out.println("\nSending Window: " + base + " to " + (end - 1));
            for (int i = base; i < end; i++) System.out.println("Sent Frame " + i);

            // 30% chance some frame in window is lost; 70% chance all OK
            if (rnd.nextInt(100) < 30) {
                int lost = base + rnd.nextInt(end - base); // [base, end-1]
                System.out.println("Receiver: Frame " + lost + " LOST");
                System.out.println("Sender: TIMEOUT → retransmit from Frame " + lost);
                base = lost; // go-back-n: retransmit from first unacked (lost) frame
            } else {
                System.out.println("Receiver: ACK all frames up to " + (end - 1));
                base = end; // slide window past acknowledged frames
            }
        }
        System.out.println("\nAll Frames Sent Successfully!");
    }
}

File: goBackN.c
#include <stdio.h>
#include <unistd.h>

#define MAX_FRAMES 5      
#define TIMEOUT 3        

void receiver(int frame, int *ack) {
    *ack = frame;  
    printf("Receiver: Acknowledged Frame %d\n", frame);
}

void sender(int total_frames, int window_size) {
    int ack = -1;    
    int base = 0;     
    int next_frame = 0;  

    while (base < total_frames) {
        
        while (next_frame < base + window_size && next_frame < total_frames) {
            printf("Sender: Sending Frame %d\n", next_frame);
            next_frame++;
        }

        printf("Sender: Waiting for acknowledgment...\n");
        receiver(base, &ack);

        if (ack == base) {
            printf("Sender: ACK for Frame %d received. Sliding window...\n", base);
            base++;  
        } else {
            printf("Sender: Timeout! Retransmitting frame %d...\n", base);
        }

        
        sleep(TIMEOUT);  
}

int main() {
    int total_frames, window_size;

    printf("Enter total number of frames: ");
    scanf("%d", &total_frames);
    printf("Enter sliding window size: ");
    scanf("%d", &window_size);

    sender(total_frames, window_size);

    return 0;
}